src/sqlfluff/rules/L003.py

```diff
@@ -100,197 +244,201 @@ def _indent_size(segments: Sequence[BaseSegment], tab_space_size: int = 4) -> in
+                    ),
+                ],
+            ),
+            *fixes,
+        ]
+
+    def _eval_new(self, context: RuleContext) -> Optional[List[LintResult]]:
+        is_last = self.is_final_segment(context)
+        memory: _MemoryDict = context.memory or _MemoryDict()
+        is_ignorable = any(
+            element.is_type(*self._ignore_types)
+            for element in context.parent_stack + (context.segment,)
+        )
+        if not context.segment.segments:
+            memory.raw_stack.append(
+                (
+                    context.segment,
+                    is_ignorable,
+                )
+            )
 
-    @staticmethod
-    def _strip_buffers(line_dict: dict) -> dict:
-        """Strip a line dict of buffers for logging."""
-        return {
-            key: line_dict[key]
-            for key in line_dict
-            if key not in ("line_buffer", "indent_buffer")
-        }
+        if not is_last:
+            return [LintResult(memory=memory)]
 
-    @classmethod
-    def _is_last_segment(
-        cls,
-        segment: BaseSegment,
-        memory: dict,
-        parent_stack: Tuple[BaseSegment, ...],
-        siblings_post: Tuple[BaseSegment, ...],
-    ) -> bool:
-        """Returns True if 'segment' is the very last node in the parse tree."""
-        if siblings_post:
-            # We have subsequent siblings. Not finished.
-            return False
-        elif parent_stack:
-            # No subsequent siblings. Our parent is finished.
-            memory["finished"].add(parent_stack[-1])
-        if segment.segments:
-            # We have children. Not finished.
-            return False
-
-        # We have no subsequent siblings or children. If all our parents are
-        # finished, the whole parse tree is finished.
-        for parent in parent_stack:
-            if parent not in memory["finished"]:
-                return False
-        return True
+        # Calculate all Line Summaries
+        line_summaries = self._process_raw_stack(
+            memory.raw_stack,
+            memory,
+            tab_space_size=self.tab_space_size,
+            templated_file=context.templated_file,
+        )
+
+        results: List[LintResult] = []
+        for line_number, summary in line_summaries.items():
+            if summary["templated_line"]:
+                continue
+            # Return fixes
+            fix = self._process_current_line(
+                line_summaries,
+                memory,
+                line_number,
+            )
+            if not fix.fixes:
+                continue
+            results.append(fix)
+
+        # Poor Man's correction of duplicate changes
+        seen = set()
+        for res in results:
+            if not res.fixes:
+                continue
+            fixes: List[LintFix] = []
+            for fix_el in res.fixes:
+                this_set = (fix_el.anchor, fix_el.edit_type)
+                if this_set in seen:
```

Comment:
In some cases in segment world (I forget when), two segments will compare equal but are not the same object. This can cause "set-like" operations to misbehave. Is that a potential concern here? I've been working around that by using `id(my_object)` or using a recently added class, `sqlfluff.core.parser.segments.base.IdentitySet`.